src/ghci/mod.rs

```diff
@@ -932,9 +993,387 @@ impl Ghci {
+                self.display_single_warning(warning).await;
+            }
+        }
+    }
+
+    /// Display a single warning with GHC-style formatting and colors.
+    async fn display_single_warning(&self, warning: &parse::GhcDiagnostic) {
+        // Format each part of the warning to match GHC's selective coloring
+        let mut parts = Vec::new();
+
+        // File path (bold, like GHC)
+        match &warning.path {
+            Some(path) => parts.push(format!(
+                "{}",
+                path.if_supports_color(Stdout, |text| text.bold())
+            )),
+            None => parts.push(format!(
+                "{}",
+                "<no location info>".if_supports_color(Stdout, |text| text.bold())
+            )),
+        }
+
+        // Position range (bold, like GHC)
+        if !warning.span.is_zero() {
+            parts.push(format!(
+                ":{}",
+                warning.span.if_supports_color(Stdout, |text| text.bold())
+            ));
+        }
+
+        // Severity with color (this is what GHC colors)
+        let severity_colored = match warning.severity {
+            parse::Severity::Warning => {
+                format!(
+                    ": {}:",
+                    "warning".if_supports_color(Stdout, |text| text.magenta())
+                )
+            }
+            parse::Severity::Error => {
+                format!(
+                    ": {}:",
+                    "error".if_supports_color(Stdout, |text| text.red())
+                )
+            }
+        };
+        parts.push(severity_colored);
+
+        // Message content with GHC-style pattern coloring
+        let colored_message = if warning.message.starts_with('\n') {
+            self.apply_ghc_message_coloring(&warning.message, warning.severity)
+        } else {
+            format!(
+                " {}",
+                self.apply_ghc_message_coloring(&warning.message, warning.severity)
+            )
+        };
+        parts.push(colored_message);
+
+        // Join all parts and display
+        let formatted = parts.join("");
+        tracing::info!("{}", formatted);
+    }
+
+    /// Apply GHC-style coloring to warning message content.
+    fn apply_ghc_message_coloring(&self, message: &str, severity: parse::Severity) -> String {
+        let lines: Vec<&str> = message.lines().collect();
+        let mut colored_lines = Vec::new();
+
+        for line in lines {
+            let colored_line = self.color_message_line(line, severity);
+            colored_lines.push(colored_line);
+        }
+
+        colored_lines.join("\n")
+    }
+
+    /// Apply colors to a single line of a warning message based on GHC patterns.
+    fn color_message_line(&self, line: &str, severity: parse::Severity) -> String {
+        // Detect different types of lines and apply appropriate coloring
+
+        // Source code lines with line numbers (e.g., "  28 | import Data.Coerce (coerce)")
```

Comment:
We'll probably want to factor this out into an actual parser and/or another file.